build/android/app/src/main/java/net/minetest/minetest/MainActivity.java

```diff
@@ -0,0 +1,137 @@
+
+public class MainActivity extends AppCompatActivity {
+	private final static int versionCode = BuildConfig.VERSION_CODE;
+	private final static int PERMISSIONS = 1;
+	private static final String[] REQUIRED_SDK_PERMISSIONS = new String[]{Manifest.permission.WRITE_EXTERNAL_STORAGE};
+	private static final String SETTINGS = "MinetestSettings";
+	private static final String TAG_VERSION_CODE = "versionCode";
+	private ProgressBar mProgressBar;
+	private TextView mTextView;
+	private SharedPreferences sharedPreferences;
+	private final BroadcastReceiver myReceiver = new BroadcastReceiver() {
+		@Override
+		public void onReceive(Context context, Intent intent) {
+			int progress = 0;
+			if (intent != null)
+				progress = intent.getIntExtra(UnzipService.ACTION_PROGRESS, 0);
+			if (progress >= 0) {
+				if (mProgressBar != null) {
+					mProgressBar.setVisibility(View.VISIBLE);
+					mProgressBar.setProgress(progress);
+				}
+				mTextView.setVisibility(View.VISIBLE);
+			} else
+				startNative();
+		}
+	};
+
+	@Override
+	public void onCreate(Bundle savedInstanceState) {
+		super.onCreate(savedInstanceState);
+		setContentView(R.layout.activity_main);
+		IntentFilter filter = new IntentFilter(UnzipService.ACTION_UPDATE);
+		registerReceiver(myReceiver, filter);
+		mProgressBar = findViewById(R.id.progressBar);
+		mTextView = findViewById(R.id.textView);
+		sharedPreferences = getSharedPreferences(SETTINGS, Context.MODE_PRIVATE);
+		if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M)
+			checkPermission();
+		else
+			checkAppVersion();
+	}
+
+	private void checkPermission() {
+		final List<String> missingPermissions = new ArrayList<>();
+		for (final String permission : REQUIRED_SDK_PERMISSIONS) {
+			final int result = ContextCompat.checkSelfPermission(this, permission);
+			if (result != PackageManager.PERMISSION_GRANTED)
+				missingPermissions.add(permission);
+		}
+		if (!missingPermissions.isEmpty()) {
+			final String[] permissions = missingPermissions
+					.toArray(new String[0]);
+			ActivityCompat.requestPermissions(this, permissions, PERMISSIONS);
+		} else {
+			final int[] grantResults = new int[REQUIRED_SDK_PERMISSIONS.length];
+			Arrays.fill(grantResults, PackageManager.PERMISSION_GRANTED);
+			onRequestPermissionsResult(PERMISSIONS, REQUIRED_SDK_PERMISSIONS, grantResults);
+		}
+	}
+
+	@Override
+	public void onRequestPermissionsResult(int requestCode, @NonNull String[] permissions, @NonNull int[] grantResults) {
+		if (requestCode == PERMISSIONS) {
+			for (int index = 0; index < permissions.length; index++) {
+				if (grantResults[index] != PackageManager.PERMISSION_GRANTED) {
+					Toast.makeText(this, R.string.not_granted, Toast.LENGTH_LONG).show();
+					finish();
+				}
+			}
+			checkAppVersion();
+		}
+	}
+
+	private void checkAppVersion() {
+		if (sharedPreferences.getInt(TAG_VERSION_CODE, 0) == versionCode)
+			startNative();
+		else
+			new CopyZipTask(this).execute(getCacheDir() + "/Minetest.zip");
```

Comment:
What are you unzipping here?

Reply:
We are unpacking in the right place. NativeActivity will only start when we make sure everything is ready!

Reply:
No, I mean: What data do you unzip here? Cache? Minetest_game? Where does this stuff come from?

Reply:
Assets folder with builtin, textures, minetest_game...